src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs

```diff
@@ -0,0 +1,108 @@
+        /// </summary>
+        private void ComputeIndices()
+        {
+            var duplicateFiles = OrderedItems
+                .Select(p => Path.GetFileName(p.EvaluatedInclude))
+                .GroupBy(file => file, StringComparer.OrdinalIgnoreCase)
+                .Where(g => g.Count() > 1)
+                .Select(g => g.Key).ToImmutableHashSet();
+
+            var index = 1;
+            
+            foreach (var item in OrderedItems)
+            {
+                var includeParts = item.EvaluatedInclude.Split(_separators, StringSplitOptions.RemoveEmptyEntries);
+                foreach (var part in includeParts)
+                {
+                    var rootedPath = duplicateFiles.Contains(part) ? _project.MakeRooted(item.EvaluatedInclude) : null;
+                    if (rootedPath != null && !_rootedOrderMap.ContainsKey(rootedPath))
+                    {
+                        _rootedOrderMap.Add(rootedPath, index++);
+                    }
+                    else if (!_displayOrderMap.ContainsKey(part))
+                    {
+                        _displayOrderMap.Add(part, index++);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Assign a display order property to items that have previously been preordered
+        /// or other (hidden) items under the project root that are not folders
+        /// </summary>
+        /// <param name="propertyContext">context for the tree item being evaluated</param>
+        /// <param name="propertyValues">mutable properties that can be changed to affect display order etc</param>
+        public void CalculatePropertyValues(
+            IProjectTreeCustomizablePropertyContext propertyContext, 
+            IProjectTreeCustomizablePropertyValues propertyValues)
+        {
+            if (propertyValues is IProjectTreeCustomizablePropertyValues2 propertyValues2)
+            {
+                var isAllowedItemType = propertyContext.ItemType != null 
+                    && _allowedItemTypes.Contains(propertyContext.ItemType);
+
+                // assign display order to folders and items that have a recognized 
+                // item type and appear in order map
+                bool hasDisplayOrder = false;
+                if (isAllowedItemType || propertyContext.IsFolder)
+                {
+                    if (_displayOrderMap.TryGetValue(propertyContext.ItemName, out var index) 
+                        || (propertyContext.Metadata.TryGetValue(FullPathProperty, out var fullPath)
+                            && _rootedOrderMap.TryGetValue(fullPath, out index)))
+                    {
+                        propertyValues2.DisplayOrder = index;
+                        hasDisplayOrder = true;
+                    }
+                }
+
+                if (!hasDisplayOrder && propertyContext.ParentNodeFlags.Contains(ProjectTreeFlags.ProjectRoot) && !propertyContext.IsFolder)
```

Comment:
If a folder isn't included where it does it appear?

Reply:
Non included folders appear at the top sorted in alphabetical order. (E.g is the hidden output folders)

---
In reply to: [149218952](https://github.com/dotnet/project-system/pull/2930#discussion_r149218952) [](ancestors = 149218952)